tensor_comprehensions/__init__.py

```diff
@@ -12,20 +12,372 @@
+        function is determined by the TC object's  mapping_options_factory which
+        can be (re)set for a TC object by calling set_mapping_options_factory.
+
+        Args:
+            entry_point: the name of the TC def to be executed
+
+            inputs: tensors that TC takes as input arguments. The inputs must
+                be passed in the order of the definition of the TC function.
+
+            outputs: tensors the TC writes to in-place. If output tensors are
+                omitted, new outputs will be allocated and returned at each
+                call.
+
+            unchecked: run in low-latency unchecked mode where failsafe size
+                and stride checks are omitted.
+
+        Returns
+            List[torch.Tensor]: the tensors returned by the TC. If outputs are
+                not specified then new allocations take place. Otherwise the TC
+                writes in-place and returns the same tensors it received as
+                outputs.
+        """
+
+        # Locally scoped implicit compilation
+        def __implicit_compile__(tc_obj: TC,
+                                 entry_point: str,
+                                 *inputs: torch.Tensor):
+            already_compiled = tc_obj.compilation_cache.is_compiled(
+                entry_point, inputs)
+
+            if already_compiled:
+                return
+
+            if not SILENT:
+                print(
+                    "TC \"{}\" was not explicitly compiled for ".format(entry_point) +
+                    "inputs of sizes:\n  {}\n".format("".join("{}, ".format(
+                        i.size().__str__()) for i in inputs)) +
+                    "....Generate implicit MappingOptions")
+
+            mapping_options = tc_obj.mapping_options_factory(
+                tc_obj.tc, entry_point, *inputs)
+
+            assert mapping_options is not None, (
+                "No options found for TC {} ".format(entry_point) +
+                "with inputs of sizes:\n  {}\n".format(
+                    "".join("{} ".format(i.size().__str__()) for i in inputs)))
+
+            # Compile best options to set the executor for the current
+            #     (entry point, inputs)
+            start = time.clock()
+            tc_obj.compilation_cache.compile(
+                entry_point, inputs, mapping_options)
+            if not SILENT:
+                print(
+                    "Done compiling TC \"{}\" (compile time: {}ms)".format(
+                        entry_point, int((time.clock() - start) * 10 ** 3)))
+
+        __implicit_compile__(self, entry_point, *inputs)
+
+        inputs = tuple(inputs) if len(inputs) > 1 else tuple(inputs, )
+        if unchecked:
+            return self.compilation_cache.unchecked_run(
+                entry_point, inputs, outputs)
+
+        return self.compilation_cache.run(entry_point, inputs, outputs)
+
+    def set_mapping_options_factory(self, mapping_options_factory):
```

Comment:
Is this setter necessary?  The field is not name-mangled anyway, and there is no additional type hinting here.  I remember commenting on the field assignment in the example code, but I was questioning the idea of changing the factory on an already-constructed object.  You can end up in a state where you have compiled some kernels with naive options (because originally using naive factory), have changed the factory to tuning, but no tuning happens for already compiled kernels.  This sounds too complex to explain, so it's probably a bad idea

Reply:
fine let users complain and decide then, I can live without it